src/analyser.cpp

```diff
@@ -1349,129 +1369,303 @@ void Analyser::AnalyserImpl::analyseModel(const ModelPtr &model)
+            if (internalVariable->mType == AnalyserInternalVariable::Type::CONSTANT) {
+                mInternalEquations.push_back(std::shared_ptr<AnalyserInternalEquation> {new AnalyserInternalEquation {internalVariable}});
+            }
+        }
 
-        std::map<VariablePtr, AnalyserEquationPtr> equationMappings;
-        std::map<AnalyserEquationPtr, AnalyserVariablePtr> variableMappings;
+        // Mark some variables as external variables, if needed.
 
-        for (const auto &internalEquation : mInternalEquations) {
-            equationMappings[internalEquation->mVariable->mVariable] = std::shared_ptr<AnalyserEquation> {new AnalyserEquation {}};
-        }
+        std::map<AnalyserInternalVariablePtr, std::vector<VariablePtr>> externalVariables;
 
-        // Make our internal variables available through our API.
+        if (!mExternalVariables.empty()) {
+            // Check whether an external variable belongs to the model being
+            // analysed, or whether it is marked as an external variable more
+            // than once through equivalence or is (equivalent to) the variable
+            // of integration.
 
-        auto stateIndex = MAX_SIZE_T;
-        auto variableIndex = MAX_SIZE_T;
+            std::map<VariablePtr, std::vector<VariablePtr>> primaryExternalVariables;
 
-        for (const auto &internalVariable : mInternalVariables) {
-            // Determine the type of the variable.
-
-            AnalyserVariable::Type type;
-
-            if (internalVariable->mType == AnalyserInternalVariable::Type::STATE) {
-                type = AnalyserVariable::Type::STATE;
-            } else if (internalVariable->mType == AnalyserInternalVariable::Type::CONSTANT) {
-                type = AnalyserVariable::Type::CONSTANT;
-            } else if ((internalVariable->mType == AnalyserInternalVariable::Type::COMPUTED_TRUE_CONSTANT)
-                       || (internalVariable->mType == AnalyserInternalVariable::Type::COMPUTED_VARIABLE_BASED_CONSTANT)) {
-                type = AnalyserVariable::Type::COMPUTED_CONSTANT;
-            } else if (internalVariable->mType == AnalyserInternalVariable::Type::ALGEBRAIC) {
-                type = AnalyserVariable::Type::ALGEBRAIC;
-            } else {
-                // This is the variable of integration, so skip it.
+            for (const auto &externalVariable : mExternalVariables) {
+                if (owningModel(externalVariable->variable()) != model) {
+                    auto issue = Issue::create();
+
+                    issue->setDescription("Variable '" + externalVariable->variable()->name()
+                                          + "' in component '" + owningComponent(externalVariable->variable())->name()
+                                          + "' is marked as an external variable, but it belongs to a different model and will therefore be ignored.");
+                    issue->setLevel(Issue::Level::MESSAGE);
```

Comment:
What is the difference between "belongs to a different model" and "doesn't belong to this model"? I mean, the current message does mention that the variable is outside the model, so...?

Reply:
The difference was earlier in the sentence where it says that the variable cannot be marked as external, but to be honest, neither are great!  Maybe saying that the component isn't a part of the model, so it and its variables cannot be included in any role of any kind?  It's mostly the language around "external variable" that has connotations which are a little confusing when you're also talking about variables outside the model.

Reply:
I am not sure we want to say that the variables of a model "cannot be included in any role of any kind". This is too vague to me. This PR introduces the concept of "external variable" and the message is related to it. So, we should make use of it. Now, if we think that the name of the concept is too confusing then we should change it.

Reply:
It's the variables **outside** the model that can't have any effect on it?

Reply:
Just in case, the situation is as follows: we have two models: `m1` and `m2`. Then, we want to analyse `m1` and before doing the analysis itself, we want to let the analyser than there is an external variable. Such an external variable must belong to `m1`. However, if we try to specify an external variable for `m1` using a variable from `m2` then we will get the above message, as illustrated in the `addExternalVariableFromDifferentModels` test.

Reply:
Considering this conversation resolved, but feel free to unresolve it, if needed.